cmd/neofs-node/reputation.go

```diff
@@ -39,6 +47,133 @@ type localTrustIterator struct {
 	epochStorage *truststorage.EpochTrustValueStorage
 }
 
+type managerBuilder struct {
+	log   *logger.Logger
+	nmSrc netmapcore.Source
+}
+
+type remoteLocalTrustProvider struct {
+	loadAddrSrc network.LocalAddressSource
+
+	clientCache interface {
+		Get(string) (apiClient.Client, error)
+	}
+}
+
+type nopReputationWriter struct{}
+
+func (nopReputationWriter) Write(reputation.Trust) error {
+	return nil
+}
+
+func (nopReputationWriter) Close() error {
+	return nil
+}
+
+type remoteLocalTrustWriter struct {
+	ctx    trustcontroller.Context
+	client apiClient.Client
+
+	buf []*reputationapi.Trust
+}
+
+func (rtp *remoteLocalTrustWriter) Write(t reputation.Trust) error {
+	val := t.Value().Float64()
+	peer := t.Peer().Bytes()
+
+	apiTrust := reputationapi.NewTrust()
+
+	apiTrust.SetValue(val)
+	apiTrust.SetPeer(reputationapi.PeerIDFromV2(peer))
+
+	rtp.buf = append(rtp.buf, apiTrust)
+
+	return nil
+}
+
+func (rtp *remoteLocalTrustWriter) Close() error {
+	prm := apiClient.SendLocalTrustPrm{}
+
+	prm.SetEpoch(rtp.ctx.Epoch())
+	prm.SetTrusts(rtp.buf)
+
+	_, err := rtp.client.SendLocalTrust(
+		rtp.ctx,
+		prm,
+	)
+
+	return err
+}
+
+type remoteLocalTrustWriterProvider struct {
+	client apiClient.Client
+}
+
+func (rtwp *remoteLocalTrustWriterProvider) InitWriter(ctx trustcontroller.Context) (trustcontroller.Writer, error) {
+	return &remoteLocalTrustWriter{
+		ctx:    ctx,
+		client: rtwp.client,
+	}, nil
+}
+
+var ErrNoServerInfoProvided = errors.New("no server info provided in InitRemove call")
+
+func (rtp *remoteLocalTrustProvider) InitRemote(srv reputationroute.ServerInfo) (trustcontroller.WriterProvider, error) {
+	if srv == nil {
+		return nil, ErrNoServerInfoProvided
+	}
+
+	addr := srv.Address()
+
+	if rtp.loadAddrSrc.LocalAddress().String() == srv.Address() {
+		// if local => return no-op writer
+		return trustcontroller.SimpleWriterProvider(new(nopReputationWriter)), nil
+	}
+
+	ipAddr, err := network.IPAddrFromMultiaddr(addr)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not convert address to IP format")
+	}
+
+	c, err := rtp.clientCache.Get(ipAddr)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not initialize API client")
+	}
+
+	return &remoteLocalTrustWriterProvider{
+		client: c,
+	}, nil
+}
+
+var ErrNoCurrentNodeFoundInNetMap = errors.New("could not find current node in NetMap")
+
+// BuildManagers sorts nodes in NetMap with HRW algorithms and trivially
+// takes the next node after the current one as the only manager.
+func (mb *managerBuilder) BuildManagers(epoch uint64, id reputationapi.PeerID) (netmapapi.Nodes, error) {
+	nm, err := mb.nmSrc.GetNetMapByEpoch(epoch)
+	if err != nil {
+		return nil, err
+	}
+
+	nodes := nm.Nodes
+
+	hrw.SortSliceByValue(nodes, epoch)
```

Comment:
We need to copy `nm.Nodes` in order to avoid violation of the network map structure, or impose an additional requirement on `managerBuilder.nmSrc` for free mutation on the structure returned by the method.

Reply:
added copying